lib/value_range_check/value_range_check_base.jinja

```diff
@@ -0,0 +1,529 @@
+        logical :: env_var_have_been_read = .false.
+
+    contains
+
+        procedure get_min_max
+
+        ! The various procedures used
+        procedure :: PreEndDeclaration
+        procedure :: PostEnd
+
+        {# Collect and declare the various procedures for the same generic interface -#}
+        {# ------------------------------------------------------------------------- -#}
+        {% set all_declares=[] -%}
+        {% set all_provides=[] -%}
+        {% for name, type, bits in ALL_TYPES %}
+        procedure :: ProvideScalar{{name}}
+          {{- all_provides.append("ProvideScalar"~name) or "" }}
+          {% for dim in ALL_DIMS %}
+        procedure :: ProvideArray{{dim}}d{{name}}
+            {{- all_provides.append("ProvideArray"~dim~"d"~name) or "" }}
+          {% endfor %}
+        {% endfor %}
+
+        {% set indent="            " %}
+        {% if GENERIC_PROVIDE %}
+        !> The generic interface for providing the value of variables:
+        generic, public :: ProvideVariable => &
+            {{all_provides|join(", &\n"+indent) }}
+        {% endif %}
+
+    end type ValueRangeCheckBaseType
+
+contains
+
+    subroutine get_min_max(this, module, region, variable_name,   &
+                           has_values, value_range)
+    !> @brief This subroutine checks if any environment variables are set
+    !! that specify minimum and/or maximum values to be checked for the
+    !! specified variable. It returns result information in
+    !! has_values (one of has_none, has_min, has_max, has_min_max).
+    !! If a minimum is specified, its value will be stored in
+    !! value_range(1). If a maximum is specified, its value will be
+    !! stores in value_range(2).
+    !! @param[in] this The instance of the ValueRangeCheckBaseType.
+    !! @param[in] module The name of the module of the instrumented
+    !!            region.
+    !! @param[in] region The name of the instrumented region.
+    !! @param[in] variable_name The name of the variable to test.
+    !! @param[out] has_values Indicates if the variable should be range tested.
+    !!             One off has_none, has_min, has_max, has_min_max
+    !! @param[out] value_range Index 1 stores the minimum value if specified,
+    !!              index 2 the maximum (if specified)
+
+        implicit none
+        class(ValueRangeCheckBaseType), intent(in), target :: this
+        character(len=*), intent(in) :: module, region, variable_name
+        integer, intent(out) :: has_values
+        real(kind=real64), dimension(2), intent(out) :: value_range
+
+        character(len=:), allocatable :: env_name, env_value, sub_str
+        ! We need a copy of the names to replace '-' with '_'
+        character(len=:), allocatable :: local_module, local_region
+        integer      :: status, value_len, index_colon, i
+
+        local_module = module
+        do i=1, len(local_module)
+            if (local_module(i:i) == "-") then
```

Comment:
I'm always shocked at what you have to do when you're not writing in Python :-) (I was about to say, "why not use 'replace'..." but then I realised.)

Reply:
Yeah, same for me :)